### ISSUE TITLE
Configuration and Lifecycle test: Is K8s ConfigMap immutable?

### ISSUE THREAD

**Issue by lixuna** [cncf/cnf-conformance#508]
[Acceptance Criteria](https://github.com/cncf/cnf-conformance/issues/508#issuecomment-740762446)
---

### Configuration and Lifecycle test: Is K8s ConfigMap immutable?

**Short description of workload test:** 
- Check if a container's ConfigMap(s) use the immutable flag

No negative points if immutable flag is not used.

MVP:
- Check every ConfigMap for each container in a K8s Workload Resource
- Scoring for single container with multiple config maps
  - All immutable - positive points (eg. 1)
  - No immutable - 0 points
  - Some immutable  - 0 points
- Scoring for single workload resource with multiple containers
  - All immutable - positive points (eg. 1)
  - No immutable - 0 points
  - Some immutable  - 0 points

**Related Best Practices:**
- K8s Workload Resource configuration should be immutable after deployment
- A CNF best practice is for its container's configuration to be immutable after deployment
- A cloud native network function best practice is that a CNF configuration should be immutable after deployment

**Test Category**
- Configuration and Lifecycle, see [USAGE.md](https://github.com/cncf/cnf-conformance/blob/master/USAGE.md#configuration-and-lifecycle-tests)

**Type of test (static or runtime)** 
-  RUNTIME

**Proof of Concept** (if available)
- [ ]  link to proof of concept of new workload test, N/A

---

**Environment set up tasks:**

**Upstream tool set up tasks: (conformance suite + upstream tools)**
- N/A

**CNF setup Tasks** 

**Sample CNF tasks:**

**Code implementation tasks:**
- [ ] 

**Documentation tasks:**
- [ ] add to [Test Categories md](https://github.com/cncf/cnf-conformance/blob/master/TEST-CATEGORIES.md) if needed 
- [ ] add to [Pseudo Code md](https://github.com/cncf/cnf-conformance/blob/master/PSEUDO-CODE.md)
- [ ] Update [USAGE md](https://github.com/cncf/cnf-conformance/blob/master/USAGE.md) if needed  
- [ ] Update [installation instructions](https://github.com/cncf/cnf-conformance#installation)  if needed  

**QA tasks**

Dev Review: 
- [ ] walk through [A/C](https://github.com/cncf/cnf-conformance/issues/82#issuecomment-594853700)
- [ ] do you get the expected result?
- [ ] if yes, 
   - [ ] move to `Needs Peer Review` column  
   - [ ] create Pull Request and follow check list
   - [ ] Assign 1 or more people for peer review
- [ ] if no, document what additional tasks will be needed

Peer review:
- [ ] walk through [A/C](https://github.com/cncf/cnf-conformance/issues/82#issuecomment-594853700)
- [ ] do you get the expected result?
- [ ] if yes, 
   - [ ] move to `Reviewer Approved` column
   - [ ] Approve pull request
- [ ] if no, 
   - [ ] document what did not go as expected, including error messages and screenshots (if possible)
   - [ ] Add comment to pull request
   - [ ] request changes to pull request

**Comment by lixuna** [comment]
### DRAFT: Acceptance Criteria for peer review (based on https://github.com/cncf/cnf-conformance/issues/82)
--
### Documentation is updated:
- [ ] On https://github.com/cncf/cnf-conformance/blob/master/USAGE.md#installable-and-upgradeable-tests, I can see:
     - [ ] the `rolling_deployment_downgrade`  test added to the USAGE.md and marked with a checkmark (or POC, depending on what is completed in this issue) 
     - [ ] a new sentence describing what the `rolling_deployment_downgrade` test does
- [ ] On https://github.com/cncf/cnf-conformance/blob/master/TEST-CATEGORIES.md#installable-and-upgradeable-tests, I can see a new sentence about `rolling_deployment_downgrade` 

###  Common steps
- [ ]  Find or create a K8s cluster to use
   - Talk with team for access to an existing cluster
   - [Kind K8s clusters](https://github.com/cncf/cnf-conformance/blob/master/KIND-INSTALL.md) will work as well
- [ ]  Set the KUBECONFIG environment variable for accessing the target K8s cluster
- [ ]  As a developer, Pull down latest version of CNF conformance test code from {FEATURE} branch
     - [ ]  Check that {FEATURE} Branch passes all A/C 
     - [ ]  Check that {FEATURE} Branch passes tests in GitHub Actions
- [ ]  As a peer reviewer, check that PR passes in GitHub Actions
     - [ ] if yes, merge PR to master
     - [ ]  if no, make a comment and request that developer make changes to pass tests 
- [ ]   As a peer reviewer, Pull down latest version of CNF conformance test code from `master` branch
- [ ]  Run `shards install`

### Spec test should pass
- [ ]  Run `crystal spec`
- [ ]  I expect to see PASSED for the tests

### Update Prerequisite checker
- [ ] N/A

### Update example CNFs cnf-conformance.yml, if needed
- [ ] continue, If there is an update to cnf-conformance.yml
- [ ] for each CNF Example
- [ ] Update the example CNF's cnf-conformance.yml to include the new changes
- [ ] Update the example CNF's README if needed
- [ ] Test the example CNF following it's README

### Update yml validator, if needed
- [ ] N/A




### I would expect the CoreDNS CNF to pass the `rolling_downgrade` test when running the individual test: (`rolling_downgrade` )
- [ ]  Follow common steps above, if you have not done so already
- [ ]  Run `./cnf-conformance setup`, if you have not done so already
- [ ]  Run  `./cnf-conformance sample_coredns_setup`
- [ ]  Run `./cnf-conformance rolling_downgrade` 
- [ ]  I expect to see the output from the test
- [ ]  I expect to see PASSED for the test

### I would expect the CoreDNS CNF to pass the `rolling_downgrade`  test when I run the category tests: (installability)
- [ ]  Follow common steps above, if you have not done so already
- [ ]   Run `./cnf-conformance setup`, if you have not done so already
- [ ]  Run  `./cnf-conformance sample_coredns_setup`
- [ ]  Run `./cnf-conformance configuration_lifecycle`
- [ ]  I expect to see the output from the test
- [ ] I expect to see PASSED for the test


### I would expect the CoreDNS CNF to pass the `rolling_downgrade`  when I run the entire test suite for workload tests (`workload`)
- [ ]   Follow common steps above, if you have not done so already
- [ ]   Run `./cnf-conformance setup`
- [ ]   Download the conformance configuration to test CoreDNS: `wget -O cnf-conformance.yml https://raw.githubusercontent.com/cncf/cnf-conformance/master/example-cnfs/coredns/cnf-conformance.yml`
- [ ]  Prepare the test suite to use the CNF by running: `./cnf-conformance cnf_setup cnf-config=./cnf-conformance.yml`
- [ ]  Run `./cnf-conformance workload`
- [ ] I expect to see the output from the platform and workload tests (all non-POC tests)
- [ ] I expect to see PASSED for the test




### Screenshots of all command and workload command output
- [ ] I can see a screenshot of rolling_downgrade command for A/C above
- [ ] I can see a screenshot of configuration_lifecycle command for A/C above
- [ ]  I can see a screenshot of workload command for A/C above

**Comment by taylor** [comment]
Level of effort in points: 0,1,2,3,5,8,13,21: 8
---

@denverwilliams - 5
@agentpoyo - 5
@wavell - 8
@williscool - 5
@nupejosh - 8 
@taylor - 5
@lixuna - 8

**Comment by williscool** [comment]
design thoughts im working on so far

1. kubectl get configmaps --all-namespaces
2. test if they are all immutable

**Comment by williscool** [comment]
SO after research I think we should only support this on k8s 1.19

there is no good way to see if a feature gate is enabled programmatically and you need to use a feature gate in 1.18

"The Kubernetes API does not provide information about feature gates enabled by the command line. "

https://stackoverflow.com/questions/50441452/check-the-feature-is-enabled-disabled-in-kubernetes

And for example what seems like the only way to get this on EKS is from the control plane server logs lol

https://docs.aws.amazon.com/eks/latest/userguide/api-server-flags.html

So its probaly a pain and no uniform.

So long story short 1.19 ^ only

**Comment by williscool** [comment]
https://www.cloudytuts.com/tutorials/kubernetes/how-to-create-immutable-configmaps-and-secrets/

**Comment by williscool** [comment]
more sure on only doing 1.19 after testing on both version.

In 1.18 with the feature gate not enabled you can set the "immutable" field  on a config map.

But its still mutable/ changable. So your configmaps are not actually immutable

**Comment by williscool** [comment]
also the test suit is using the latest kind nodes which are k8s 1.19

https://github.com/kubernetes-sigs/kind/releases/tag/v0.9.0

https://github.com/cncf/cnf-conformance/runs/1684664568?check_suite_focus=true

**Comment by williscool** [comment]
Acutally... now that I think about it we could feature test by installing a test config map and see if we get the error response


```
The ConfigMap "myapp" is invalid: data: Forbidden: field is immutable when `immutable` is set
```

then say "immucateble configmaps not enabled for more infor check. "

and link to 

https://kubernetes.io/docs/concepts/configuration/configmap/#configmap-immutable

I think I like that best

**Comment by wavell** [comment]
Seems to be having intermittent failures:

https://github.com/cncf/cnf-conformance/runs/1787246041?check_suite_focus=true

the pr pass the first time through, but failed on a rerun (branches pulled from master are failing the specs)

The problem seems to be here:

https://github.com/cncf/cnf-conformance/blob/master/src/tasks/workload/configuration_lifecycle.cr#L475

There is a comparison to make sure *all* config maps have an immutable flag, but the comparison should be *only on the config maps that were installed by the cnf*.  This could be done using a helm template export.

For instance, the ca cert config map isn't installed with the cnf, and it is not immutable, but will force the test to fail:
![image](https://user-images.githubusercontent.com/29548/106200173-33ede600-617c-11eb-93f3-667266d95670.png)


@williscool

**Comment by agentpoyo** [comment]
Seeing the following error:

```
./cnf-conformance cnf_setup cnf-config=./sample-cnfs/sample_immutable_configmap_some/cnf-conformance.yml
2021-02-12T22:38:05.788848Z   INFO - CNFManager::Points::Results.file
2021-02-12T22:38:05.789029Z   INFO - file exists?:false
Successfully created directories for cnf-conformance
Error: YAML parse error on coredns/templates/configmap-test-not-immutable.yaml: error converting YAML to JSON: yaml: line 2: mapping values are not allowed in this context
Error: YAML parse error on coredns/templates/configmap-test-not-immutable.yaml: error converting YAML to JSON: yaml: line 2: mapping values are not allowed in this context
```

**Comment by agentpoyo** [comment]
I took the configmap-test-not-immutable.yml template, ran through a YAML Lint and it spit out the following to use:

```
--- 
apiVersion: v1
data: 
  api.server: "https://example.com"
kind: ConfigMap
metadata: 
  name: myapp
```

The original file was or is:
```
  apiVersion: v1
  kind: ConfigMap
  metadata:
    name: myapp
  data:
    api.server: https://example.com
```

After manually updating, `./cnf-conformance cnf_setup cnf-config=./sample-cnfs/sample_immutable_configmap_some/cnf-conformance.yml` was successful but then running `./cnf-conformance immutable_configmap` still showed PASSED but with the following output:

```
$ LOG_LEVEL=error ./cnf-conformance immutable_configmap
2021-02-12T22:54:22.739231Z   INFO - CNFManager::Points::Results.file
2021-02-12T22:54:22.739483Z   INFO - file exists?:false
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
The ConfigMap "myapp" is invalid: data: Forbidden: field is immutable when `immutable` is set
✔️  PASSED: All volume or container mounted configmaps immutable ⚖️
```

@williscool

**Comment by agentpoyo** [comment]
![508_immutable_config_some](https://user-images.githubusercontent.com/14348948/107831646-23c92f80-6d54-11eb-9ca5-e50cc91b3841.png)

**Comment by agentpoyo** [comment]
![508_immutable_config_coredns](https://user-images.githubusercontent.com/14348948/107831651-27f54d00-6d54-11eb-9f4f-9d5863c220ea.png)

**Comment by lixuna** [comment]
@agentpoyo @williscool What is remaining on this issue?

**Comment by agentpoyo** [comment]
> @agentpoyo @williscool What is remaining on this issue?

https://github.com/cncf/cnf-conformance/issues/508#issuecomment-778501963

**Comment by nupejosh** [comment]
![Screen Shot 2021-03-03 at 3 51 58 PM](https://user-images.githubusercontent.com/3164983/109877255-6639aa00-7c38-11eb-9340-6e8d908c928c.png)
